itou/approvals/management/commands/migrate_poleemploi_notification_log.py

```diff
@@ -0,0 +1,135 @@
+        paginator = Paginator(self.queryset, DEFAULT_PAGINATION_SIZE)
+        try:
+            self.stdout.write(f"migration pagination total_range={paginator.page_range}")
+            for page_number in paginator.page_range:
+                self.stdout.write(f"migration pagination page number={page_number}")
+                page = paginator.page(page_number)
+                for obj in page.object_list:
+                    self._migrate_object(obj, wet_run)
+        except KeyboardInterrupt:
+            self.stdout.write("! keyboard interrupt, exiting.")
+
+    def _migrate_object(self, obj, wet_run=False, at=None):
+        raise NotImplementedError("please override me")
+
+
+class Command(PaginatedMigrationCommand):
+    """This is a one-off command to "migrate" the data contained in the
+    JobApplicationPoleEmploiNotificationLog table to the new columns in the Approval table.
+
+    Why I don't do this in a data migration ? Because our migrations are run sequentially,
+    during the deployment, by Clever Cloud and should not take too long. I'm expecting
+    this migration to be actually quite long since some Python is involved.
+
+    Also this gives us the opportunity to run the migration and stop it whenever we feel like it.
+    """
+
+    help = "Migrates the notification data from its dedicated log table to the approvals table."
+
+    queryset = (
+        JobApplication.objects.exclude(jobapplicationpoleemploinotificationlog__isnull=True)
+        .filter(approval__pe_notification_status=approvals_enums.PEApiNotificationStatus.PENDING)
+        .select_related("approval")
+        .prefetch_related("jobapplicationpoleemploinotificationlog_set")
+    )
+
+    def _migrate_object(self, obj, wet_run=False, at=None):
+        job_application = obj
+        approval = job_application.approval
+
+        if not approval:
+            self.stdout.write(f"> removing logs for job_application={job_application} without approval")
+            if wet_run:
+                job_application.jobapplicationpoleemploinotificationlog_set.all().delete()
+            return
+
+        if job_application.jobapplicationpoleemploinotificationlog_set.filter(
+            status=JobApplicationPoleEmploiNotificationLog.STATUS_OK
+        ).exists():
+            self.stdout.write(
+                f"> handling job_application={job_application} approval={approval} as notification success."
+            )
+            if wet_run:
+                approval.pe_save_success(at)
+        else:
+            for log in job_application.jobapplicationpoleemploinotificationlog_set.order_by("-created_at").all():
+                exit_code = exit_code_from_details(log.details)
+                if exit_code and approval:
+                    endpoint = approvals_enums.PEApiEndpoint.MISE_A_JOUR_PASS_IAE
+                    if log.status == JobApplicationPoleEmploiNotificationLog.STATUS_FAIL_SEARCH_INDIVIDUAL:
```

Comment:
Je ne comprends pas l'identation de ce `if`. 
Soit  `log.status == JobApplicationPoleEmploiNotificationLog.STATUS_FAIL_SEARCH_INDIVIDUAL` et peut importe `exit_code` et `approval`. 
Soit` exit_code and approval` et peu importe `log.status`

Reply:
hmm, non, je pense que une info a été loupée.
En gros:
- `exit_code_from_details` retourne quelque chose si et seulement si on trouve un "vrai" exit code, donc que l'on est dans le cas d'une erreur non récupérable.
- il faut également que on ait une `approval` sur l'objet parce que actuellement ce n'est pas obligatoire (!) je pense que Clément avait tenté des trucs bizarres à des moments; on a donc des logs sur des `JobApplication` sans PASS, va savoir.
- donc si j'ai un PASS et un code d'erreur, c'est que je suis dans le cas d'une "vraie" erreur et donc je veux l'endpoint (MISE A JOUR ou RECHERCHE) et ça, je le devine en lisant le `status` du `JobApplicationPoleEmploiNotificationLog` qui donne plus ou moins cette info. Sinon je serais obligé de "choisir" arbitrairement quel etait l'endpoint concerné, mais j'essaie de faire au plus proche dans cette migration.
Tu vois l'idée ?

Reply:
Ça fait sens. Merci pour tes précisions.

Reply:
oui bah en meme temps, comme quoi le systeme de logs etait quand meme pas facile a suivre actuellement tu vois ^^ :sweat_smile: